src/web/js/output-ui.js

```diff
@@ -1256,57 +945,38 @@
-                 handleMouseEvents: false, atomic: true});
-              if(!editors[key.l.source]) {
-                var styles = document.getElementById("highlight-styles").sheet;
-                styles.insertRule(
-                  ((context === undefined) ? "" : "#main[data-highlights=" + context + "]")
-                    + " ." + locKey + " { " + (!!key.c ? "background-color:" + key.c : "")
-                    + ";border-bottom: 2px hsla(0, 0%, 0%,.5) solid;}",styles.cssRules.length);
-              };
-              if(!key.c)
-                return;
-              var updated = false;
-              s.editor.on("update", function() {
-                if(updated) return;
-                updated = true;
-                $(s.wrapper).find(".highlight." + cmlocToCSSClass(key.l)).on("mouseenter", function() {
-                  hintLoc(runtime, editors, srcloc, key.pl);
-                  $("."+cmlocToCSSClass(key.l)).css("animation", "pulse 0.4s infinite alternate");
-                });
-                $(s.wrapper).find(".highlight." + cmlocToCSSClass(key.l)).on("click", function() {
-                  emphasizeLine(editors, key.l);
-                  gotoLoc(runtime, editors, srcloc, key.pl);
-                });
-                $(s.wrapper).find(".highlight." + cmlocToCSSClass(key.l)).on("mouseleave", function() {
-                  unhintLoc(runtime, editors, srcloc, key.pl);
-                  $("."+cmlocToCSSClass(key.l)).css("animation", "");
-                });
-              });
-            });
+
+        messagePositions.forEach(function(positions, color) {
+          var snippetPositions = [];
+          positions.forEach(function (position) {
+            Array.prototype.push.apply(snippetPositions,
+              (snippets.get(position.source) || []).
+                map(function (doc) {
+                  return new Position(doc, position.source, position.from, position.to);
+                }));
           });
+          Array.prototype.push.apply(positions, snippetPositions);
+          allHighlightPositions.set(color, positions);
         });
 
-        if(context === undefined) 
-          rendering.addClass("highlights-active");
-        else if(context != undefined) {
-          rendering.bind('toggleHighlight',function() {
-            $(".highlights-active").removeClass("highlights-active");
-            document.getElementById("main").dataset.highlights = context;
-            rendering.addClass("highlights-active");
+        messageAnchors.forEach(function(anchors, color) {
+          allHighlightAnchors.set(color, anchors);
+        });
+
+        rendering.bind('toggleHighlight',function() {
+          colorsHighlighted.forEach(function(color) {
+            unhighlight(color);
           });
-        }
+          colorsEmphasized.forEach(function(color) {
+            demphasize(color);
+          });
+          if(sessionStorage.getItem('highlight-eagerness') != 'lazy') {
```

Comment:
Ditto another comment about using `sessionStorage` in the middle of other code – probably better to parameterize over this.